File: app/src/main/java/br/com/menuvem/lojista/domain/model/ProdutoComponente.kt
package br.com.menuvem.lojista.domain.model

/**
 * Vínculo de um [Componente] a um [Produto], com o multiplicador aplicado.
 *
 * @param multiplicador Fração do componente usada no produto. Ex.: sabor único
 *        numa pizza = 1.0 (inteiro); 2 sabores = 0.5 (metade); 3 sabores = 1/3.
 *        Massa/base e embalagem normalmente usam 1.0.
 */
data class ProdutoComponente(
    val id: Long = 0,
    val produtoId: Long,
    val componenteId: Long,
    val multiplicador: Double = 1.0
)

File: lib/domain/model/produto_componente.dart
/// Vínculo de um Componente a uma [Porcao], com o multiplicador aplicado.
///
/// [componenteId]: qual componente (independente do tamanho).
/// [tamanhoComponenteId]: qual tamanho **desse** componente foi aplicado —
/// todo componente tem ao menos um tamanho ("Único"), então este campo
/// nunca fica em aberto.
/// [multiplicador]: fração do componente usada na porção. Ex.: sabor único
/// numa pizza = 1.0 (inteiro); 2 sabores = 0.5 (metade); 3 sabores = 1/3.
/// Massa/base e embalagem normalmente usam 1.0.
class ProdutoComponente {
  final int id;
  final int porcaoId;
  final int componenteId;
  final int tamanhoComponenteId;
  final double multiplicador;

  const ProdutoComponente({
    this.id = 0,
    required this.porcaoId,
    required this.componenteId,
    required this.tamanhoComponenteId,
    this.multiplicador = 1.0,
  });

  ProdutoComponente copyWith({
    int? id,
    int? porcaoId,
    int? componenteId,
    int? tamanhoComponenteId,
    double? multiplicador,
  }) {
    return ProdutoComponente(
      id: id ?? this.id,
      porcaoId: porcaoId ?? this.porcaoId,
      componenteId: componenteId ?? this.componenteId,
      tamanhoComponenteId: tamanhoComponenteId ?? this.tamanhoComponenteId,
      multiplicador: multiplicador ?? this.multiplicador,
    );
  }
}
